RIPS/rip-9997.md

```diff
@@ -0,0 +1,198 @@
+```
+BUNDLE_TRANSACTION_SUBTYPE = x
+```
+
+These transactions are completely identical to regular RIP-7560 transactions with the exceptions of being
+composable into bundles.
+
+It is important for the wallets to explicitly opt into this feature, as some Smart Contract Accounts
+may be created in a way that relies on atomicity between validation and execution phases.
+
+### All validation state changes apply before all execution ones
+
+Filling a block with AA transactions must not be a challenge for the block builder.
+However, if each transaction during its execution can alter any state that affects the validity of another transaction
+in the mempool, the block builder will be forced to revalidate all transactions in the mempool after each inclusion.
+
+We mitigate that by applying all changes in all the validation frames of a sequence of AA transactions first
+and all execution frames apply immediately after that.
+
+In theory, the validation frames can also invalidate each other, but we define ways to prevent that by applying
+certain rules for the mempool transactions in ERC-7562.
+
+A builder that chooses not to enforce the rules from ERC-7562 **must** take care to re-validate each transaction
+against the mid-block state at the position where it is being included into a block.
+Otherwise, the resulting block is likely to end up being invalid.
+
+### Block structure diagram
+
+Here is a visual representation of a block that contains multiple Account Abstraction Transactions.
+The validation parts of AA transactions are executed as separate transactions,
+but are not represented as separate transactions in the block data.
+
+![](../assets/rip-7560/block_overview.png)
+*The structure of a block containing multiple Native Account Abstraction Transactions*
+
+### Validation state change virtual transactions
+
+The validation frames of the AA_TX_TYPE transaction are represented as individual virtual transactions by the clients.
+They are assigned their own sequential `transactionIndex`, and their `transactionHash` is defined as
+(`AA_TX_TYPE transaction hash + 1`).
+
+All block-related RPC methods, like `eth_getBlockByHash` and `eth_getBlockByNumber`, must include these virtual
+transactions as part of the `transactions` field and include validation in the block transaction count.
+
+All transaction-related RPC methods, like `eth_getTransactionByHash` and `eth_getTransactionReceipt`, must
+accept the virtual transaction hash as input and return the details calculated as if the validation was a
+separate transaction.
+
+There is a number of behaviours that define transaction-wide effects in Ethereum.
+This list includes, but is not limited to:
+
+* Tracking `accessed_addresses`
+* [EIP-1283](https://eips.ethereum.org/EIPS/eip-1283) Gas metering for SSTORE
+* [EIP-1153](https://eips.ethereum.org/EIPS/eip-1153) Transient storage opcodes
+
+Any such behaviour has separate effects in the "Validation Virtual Transaction" and "Execution Transaction".
```

Comment:
You mean TSTOREs are lost between validation and execution, and warming fees for accounts and slots have to be paid twice during the transaction?  If so, why?

Reply:
yes, TSTORE is lost.
The reason we run validation in sequence, and then all execution. it is going to be extremely hard to keep TSTOREs and pass them to the respective execution.
We do want TSTORE to be independent between different executions, and the side-effect is that validation frames  also have a separate TSTORE namespace.

Reply:
I think that ignoring TSTORE benefits is not the way to go. The tradeoff between added complexity in the RIP/implementation vs higher gas costs permanently for many transactions, should always go in favor of lowering gas costs...

Reply:
its not a higher gas cost: its higher complexity, requiring to keep a "transient" storage in-memory between transactions (frames) in the same block.
Also, it is not a shared storage (like the global one): it is a transient storage of specific userop.
also: what would you do if the same account (sender) is used between different transactions? would you share the transient storage between also those same-account transactions ?

Lastly: if you want to add cross-frame transient storage, then we can add it later in a separate RIP. I think it has its own can of worms to handle.

Reply:
@drortirosh @yoavw 
This is a major issue and it seems like there is still a disagreement about what to do with the TSTORE for now. I will go with the results of the vote in the "Open Questions" doc but we probably should discuss this further.